### PR TITLE
GUI

### PR DESCRIPTION
**Pull request by ll-nick**
(Actual draft PRs are a paid feature so the title will have to do)

Adds @orzechow fancy GUI web app to the repo and the demo while replacing the ROS based communication with something more generic, probably web sockets.

### CODE REVIEW

**Comment by orzechow** [comment]
Alright, the GUI is ready for review :tada: 

I moved all CMake cleanup stuff into #68 to separate concerns a bit.


https://github.com/user-attachments/assets/96c874cc-02a5-4585-8d79-1d1ea6581cfc

File: demo/src/main.cpp [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
 using namespace utils;
+using namespace arbitration_graphs;
 
 int main() {
     try {
         PacmanWrapper demo;
         PacmanAgent agent(demo.game());
 
+        gui::WebServer server(8080, true);
+        server.loglevel(crow::LogLevel::Info);
```
Comment (ll-nick) [inline demo/src/main.cpp:21]: I'd set this to `Warning`, right? There's quite a bit of spam in the terminal right now that would not really be necessary.

**Review by ll-nick — COMMENTED** [review]
Very, very nice, thanks for all the hard work!

This is a very nice and important feature and it's integrated really well. Most complaints I have are minor (and since I'm neither a JS guy nor the CMake expert between the two of us you probably know better anyway ;-) ).
The one thing I'd  wish for if it is feasible without too much overhead would be the ability to configure the port of the vue app.

Also, you should update this line in the current README:
```md
We will shortly add an [arbitration graph GUI](https://github.com/KIT-MRT/arbitration_graphs/pull/10) and a [tutorial](https://github.com/KIT-MRT/arbitration_graphs/pull/51) based on this demo – stay tuned! 
```

I opened the PR, so you'll have to approve it yourself.

File: gui/htmlroot/arbitrator_option.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                // This component is a behavior option → rounded edges
+                return 10;
+            } else {
+                // This component is an arbitrator → square edges
+                return 0;
+            }
+        },
+        optionRy: function () {
+            return this.optionRx();
+        },
```
[the remark this answers is not in the document]
Comment (orzechow) [reply gui/htmlroot/arbitrator_option.js:123]: Weird, which Firefox version do you use? :thinking: 
The video above has been recorded from my Firefox 131.0.3

File: gui/htmlroot/index.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            showButtons: true,
+            showStatus: true,
+            hoverTimeout: null,
+            hoverAnimationTimeout: null,
+            animateHoverBar: false
+        }
+    },
+    methods: {
+        connectToWebSocket() {
+            const wsUrl = 'ws://localhost:8080/status';
```
[the remark this answers is not in the document]
Comment (orzechow) [reply gui/htmlroot/index.js:27]: That's a really good point. Let me look into it

File: gui/include/arbitration_graphs/gui/web_server.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        std::lock_guard<std::mutex> guard(connections_mutex_);
+        for (auto* conn : connections_) {
+            conn->send_text(message);
+        }
+
+        CROW_LOG_DEBUG << "Message sent to all clients: " << message;
+    }
+
+    crow::SimpleApp loglevel(crow::LogLevel level) {
+        return app_.loglevel(level);
```
[the remark this answers is not in the document]
Comment (orzechow) [reply gui/include/arbitration_graphs/gui/web_server.hpp:75]: Thought about it too – for me it's not intuitive, but I preferred to follow the crow interface :shrug:

[the remark this answers is not in the document]
Comment (orzechow) [reply gui/include/arbitration_graphs/gui/web_server.hpp:75]: Probably that comes from the crow approach of allowing stuff like

`app_.port(port_).multithreaded().run_async();`

Same for `loglevel()`:

`app_.loglevel(crow::LogLevel::Info).port(port_).multithreaded().run_async()`

[the remark this answers is not in the document]
Comment (orzechow) [reply gui/include/arbitration_graphs/gui/web_server.hpp]: As we actually don't want to expose the private app object (or when, we should add a proper get function instead), I changed to `void`: 16260a6

**Comment by orzechow** [comment]
Rebased onto main

File: gui/htmlroot/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (orzechow) [reply gui/htmlroot/index.js]: Alright, I first went for serving a config.json via Crow like this:

```cpp
CROW_ROUTE(app_, "/config.json")
([this]() { return crow::json::wvalue{{"websocket_port", port_}}; });
```

But in the end, it even easier by using the app port from the current URI location: 786a76f

```javascript
const wsUrl = 'ws://localhost:' + location.port + '/status';
```

:sweat_smile:

**Comment by orzechow** [comment]
Alright – now, that we don't compile the static files path into an object file, but pass it as pre-compiled header (5bc50c9),
I gave it a shot to turn the GUI into a header-only lib as well: 3f1fa33 :tada: 

@ll-nick Please test this thoroughly on your machine (remember to disable or even prune docker caches)

Also, I figured that the demo tests didn't compile anymore: 8ed4e40

As soon as you're happy with these changes, we can merge :smile:

File: gui/htmlroot/arbitrator_option.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ll-nick) [reply gui/htmlroot/arbitrator_option.js:123]: Hm, weird. I use 132.0.

File: gui/htmlroot/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ll-nick) [reply gui/htmlroot/index.js]: Ah yes, very clever. Thanks, that's much nicer :ok_hand:

File: gui/htmlroot/arbitrator_option.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (orzechow) [reply gui/htmlroot/arbitrator_option.js:123]: Hm, after updating to 132.0.1 I still don't see this issue…

**Comment by ll-nick** [comment]
Alrighty, I tested everything I believe
- Building the library docker
- Building the devel docker and run the unit tests
- Build and run the demo
- Build the tutorial and run the unit tests

Everything works flawlessly except for the colors of the GUI being off for me. If it works for you, I'd merge anyway and look into this in a separate issue or something. Maybe I also just need to delete some browser cache or similar.

**Review by orzechow — APPROVED** [review]
Approving my own stuff feels weird :see_no_evil:

**Comment by ll-nick** [comment]
> Approving my own stuff feels weird 🙈

image

**Pull request merged**